src/coreclr/src/vm/pgo.cpp

```diff
@@ -342,6 +479,166 @@ HRESULT PgoManager::getMethodBlockCounts(MethodDesc* pMD, unsigned ilSize, UINT3
     return E_NOTIMPL;
 }
 
+CORINFO_CLASS_HANDLE PgoManager::getLikelyClass(MethodDesc* pMD, unsigned ilSize, unsigned ilOffset, UINT32* pLikelihood, UINT32* pNumberOfClasses)
+{
+    *pLikelihood = 0;
+    *pNumberOfClasses = 0;
+
+    // Bail if there's no profile data.
+    //
+    if (s_PgoData == NULL)
+    {
+        return NULL;
+    }
+
+    // See if we can find profile data for this method in the profile buffer.
+    //
+    const unsigned maxIndex = s_PgoIndex;
+    const unsigned token    = pMD->IsDynamicMethod() ? 0 : pMD->GetMemberDef();
+    const unsigned hash     = pMD->GetStableHash();
+
+    unsigned index = 0;
+    unsigned methodsChecked = 0;
+
+    while (index < maxIndex)
+    {
+        // The first two "records" of each entry are actually header data
+        // to identify the method.
+        //
+        Header* const header = (Header*)&s_PgoData[index];
+
+        // Sanity check that header data looks reasonable. If not, just
+        // fail the lookup.
+        //
+        if ((header->recordCount < MIN_RECORD_COUNT) || (header->recordCount > MAX_RECORD_COUNT))
+        {
+            break;
+        }
+
+        // See if the header info matches the current method.
+        //
+        if ((header->token == token) && (header->hash == hash) && (header->ilSize == ilSize))
+        {
+            // Yep, found data. See if there is a suitable class profile.
+            //
+            // This bit is currently somewhat hacky ... we scan the records, the count records come
+            // first and are in increasing IL offset order. Class profiles have inverted IL offsets
+            // so when we find an offset with high bit set, it's going to be an class profile.
+            //
+            unsigned countILOffset = 0;
+            unsigned j = 2;
+
+            // Skip past all the count entries
+            //
+            while (j < header->recordCount)
+            {
+                if ((s_PgoData[index + j].ILOffset & ICorJitInfo::ClassProfile::CLASS_FLAG) != 0)
+                {
+                    break;
+                }
+
+                countILOffset = s_PgoData[index + j].ILOffset;
+                j++;
+            }
+
+            // Now we're in the "class profile" portion of the slab for this method.
+            // Look for the one that has the right IL offset.
+            //
+            while (j < header->recordCount)
+            {
+                const ICorJitInfo::ClassProfile* const classProfile = (ICorJitInfo::ClassProfile*)&s_PgoData[index + j];
+
+                if ((classProfile->ILOffset & ICorJitInfo::ClassProfile::OFFSET_MASK) != ilOffset)
+                {
+                    // Need to make sure this is even divisor
+                    //
+                    j += sizeof(ICorJitInfo::ClassProfile) / sizeof(ICorJitInfo::BlockCounts);
+                    continue;
```

Comment:
I don't understand why we need to make sure this is an even divisor?

Reply:
Because the underlying data is allocated in BlockCount increments, we need to make sure ClassProfile is a multiple, so that this update to `j` gets us to the next ClassProfile record.